src/uwtools/tests/config/formats/test_base.py

```diff
@@ -0,0 +1,149 @@
+    print(config)
+    assert yaml.safe_load(capsys.readouterr().out)["foo"] == 88
+
+
+def test__load_paths(config, tmp_path):
+    paths = (tmp_path / fn for fn in ("f1", "f2"))
+    for path in paths:
+        with open(path, "w", encoding="utf-8") as f:
+            yaml.dump({path.name: "defined"}, f)
+    cfg = config._load_paths(config_files=paths)
+    for path in paths:
+        assert cfg[path.name] == "present"
+
+
+def test_characterize_values(config):
+    values = {1: "", 2: None, 3: "{{ n }}", 4: {"a": 88}, 5: [{"b": 99}], 6: "string"}
+    complete, empty, template = config.characterize_values(values=values, parent="p")
+    assert complete == ["    p4", "    p4.a", "    pb", "    p5", "    p6"]
+    assert empty == ["    p1", "    p2"]
+    assert template == ["    p3: {{ n }}"]
+
+
+@pytest.mark.parametrize("fmt", [FORMAT.ini, FORMAT.nml, FORMAT.yaml])
+def test_compare_config(caplog, fmt, salad_base):
+    """
+    Compare two config objects.
+    """
+    log.setLevel(logging.INFO)
+    cfgobj = tools.format_to_config(fmt)(fixture_path(f"simple.{fmt}"))
+    if fmt == FORMAT.ini:
+        salad_base["salad"]["how_many"] = "12"  # str "12" (not int 12) for ini
+    assert cfgobj.compare_config(salad_base) is True
+    # Expect no differences:
+    assert not caplog.records
+    caplog.clear()
+    # Create differences in base dict:
+    salad_base["salad"]["dressing"] = "italian"
+    salad_base["salad"]["size"] = "large"
+    del salad_base["salad"]["how_many"]
+    assert not cfgobj.compare_config(cfgobj, salad_base)
+    assert not cfgobj.compare_config(salad_base)
+    # Expect to see the following differences logged:
+    for msg in [
+        "salad:        how_many:  - 12 + None",
+        "salad:        dressing:  - balsamic + italian",
+        "salad:            size:  - None + large",
+    ]:
+        assert logged(caplog, msg)
+
+
+def test_depth(config):
+    assert config.depth == 1
+
+
+def test_dereference(caplog, config):
+    # Test demonstrates that:
+    #   - Config dereferencing uses environment variables.
+    #   - Initially-unrenderable values do not cause errors.
+    #   - Initially-unrenderable values may be rendered via iteration.
+    #   - Finally-unrenderable values do not cause errors and are returned unmodified.
+    log.setLevel(logging.DEBUG)
+    config.data.update({"a": "{{ b.c + 11 }}", "b": {"c": "{{ N | int + 11 }}"}, "d": "{{ X }}"})
+    with patch.dict(os.environ, {"N": "55"}, clear=True):
+        config.dereference()
+    for excerpt in [
+        "'a': '{{ b.c + 11 }}', 'b': {'c': '{{ N | int + 11 }}'}, 'd': '{{ X }}'",
+        "'a': '{{ b.c + 11 }}', 'b': {'c': 66}, 'd': '{{ X }}'",
+        "'a': 77, 'b': {'c': 66}, 'd': '{{ X }}'",
+    ]:
+        assert regex_logged(caplog, excerpt)
+    assert config == {"foo": 88, "a": 77, "b": {"c": 66}, "d": "{{ X }}"}
+
+
+# NB: Need direct test for parse_include().
+
+# NB: Need direct test for update_values().
```

Comment:
These two methods need direct tests. I tried, but failed, to write them. I think I need a refresher on how the `!INCLUDE` directive is meant to behave. I'm going to check the user guide to see if it's described, but everything I tried didn't work as I expected it to, so I'm confused. It's ok for now: The code is transitively covered by other tests, but it should have direct tests.